# Fase 4 — Treinamento

Treina o `HotelReviewClassifier` (XLM-RoBERTa multi-task) e salva o melhor checkpoint em `models/saved/`.

## Fluxo do notebook

| Célula | O que faz |
|---|---|
| 1. Setup | Seed, device (CPU/GPU), paths |
| 2. Dataset e DataLoaders | Carrega CSV, divide treino/val 80/20, cria batches |
| 3. Modelo, Criterion e Optimizer | Instancia o classificador, as losses e o AdamW |
| 4. Loop de treino | Roda N epochs, avalia F1 e MAE, salva melhor checkpoint |
| 5. Resultado | Confirma o checkpoint salvo e as métricas finais |

> **Metas:** F1 macro ≥ 0.80 (sentimento) · MAE ≤ 0.20 normalizado / ≤ 0.8 na escala 1–5 (rating)

---

## 1. Setup

Define seed de reprodutibilidade e detecta o device disponível (GPU via CUDA ou CPU). Com GPU, cada epoch leva ~54 min na RTX 3050; na CPU, ~194 min.

In [1]:
import sys
sys.path.append('..')

import torch
import random
import numpy as np

# seed de reprodutibilidade
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)

DEVICE   = torch.device("cuda" if torch.cuda.is_available() else "cpu")
CSV_PATH = "../data/processed/reviews_labeled.csv"
SAVE_DIR = "../models/saved"

print(f"Device: {DEVICE}")

Device: cuda


## 2. Dataset e DataLoaders

Carrega `reviews_labeled.csv` (76k reviews) e divide em treino (80%) e validação (20%) de forma reprodutível. O `DataLoader` serve os dados em batches de 32 reviews — o modelo nunca vê o dataset inteiro de uma vez, só um batch por vez.

- `shuffle=True` no treino: embaralha a cada epoch para evitar que o modelo aprenda a ordem dos dados  
- `shuffle=False` na validação: ordem não importa, só as métricas

In [2]:
from torch.utils.data import DataLoader, random_split
from src.data.dataset import ReviewDataset

BATCH_SIZE = 32
VAL_RATIO  = 0.2

dataset = ReviewDataset(CSV_PATH)

val_size   = int(len(dataset) * VAL_RATIO)
train_size = len(dataset) - val_size
train_ds, val_ds = random_split(dataset, [train_size, val_size],
                                generator=torch.Generator().manual_seed(SEED))

train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, shuffle=True)
val_loader   = DataLoader(val_ds,   batch_size=BATCH_SIZE, shuffle=False)

print(f"Treino:    {train_size:,} reviews  ({len(train_loader)} batches)")
print(f"Validacao: {val_size:,} reviews  ({len(val_loader)} batches)")

c:\Users\João Marcelo\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Treino:    61,060 reviews  (1909 batches)
Validacao: 15,264 reviews  (477 batches)


## 3. Modelo, Criterion e Optimizer

- **`HotelReviewClassifier`**: backbone XLM-RoBERTa + 4 cabeças de classificação (~278M parâmetros)
- **`MultiTaskLoss`**: combina CrossEntropy (sentimento, prioridade), BCEWithLogitsLoss (categoria) e MSELoss (rating) com pesos por tarefa. Inclui class weights para compensar o desbalanceamento de sentimento (71% positivo / 9% negativo)
- **`AdamW`**: otimizador padrão para fine-tuning de transformers. `lr=2e-5` é pequeno para não destruir os pesos pré-treinados. `weight_decay=0.01` evita overfitting
- **`OneCycleLR`**: scheduler que varia o learning rate ao longo do treino (sobe até `max_lr` na metade, desce depois) — converge mais rápido que LR fixo

In [ ]:
from torch.optim import AdamW
from torch.optim.lr_scheduler import OneCycleLR
from src.models.classifier import HotelReviewClassifier
from src.models.trainer import MultiTaskLoss

EPOCHS   = 5
LR       = 2e-5

model     = HotelReviewClassifier().to(DEVICE)
criterion = MultiTaskLoss().to(DEVICE)
optimizer = AdamW(model.parameters(), lr=LR, weight_decay=0.01)
scheduler = OneCycleLR(optimizer, max_lr=LR,
                       steps_per_epoch=len(train_loader),
                       epochs=EPOCHS)

print(f"Parametros treinaveis: {sum(p.numel() for p in model.parameters() if p.requires_grad):,}")

## 4. Loop de Treino

A cada epoch:
1. `train_epoch` — processa todos os 1909 batches de treino: forward → loss → backward → optimizer step
2. `eval_epoch` — avalia nos 477 batches de validação sem atualizar pesos: coleta predições e calcula F1 macro e MAE
3. Se o F1 melhorou, salva checkpoint em `models/saved/best_model.pt`

O log por epoch permite identificar overfitting: se a loss de treino cai mas o F1 de validação trava, o modelo está memorizando os dados de treino.

In [ ]:
from src.models.trainer import train_epoch, eval_epoch

best_f1 = 0.0

for epoch in range(1, EPOCHS + 1):
    train_losses = train_epoch(model, train_loader, optimizer, criterion, DEVICE, scheduler)
    val_metrics  = eval_epoch(model, val_loader, DEVICE)

    print(
        f"Epoch {epoch}/{EPOCHS} | "
        f"loss sentiment={train_losses['sentiment']:.4f} "
        f"category={train_losses['category']:.4f} "
        f"rating={train_losses['rating']:.4f} | "
        f"F1={val_metrics['f1_macro']:.4f} "
        f"MAE={val_metrics['mae']:.4f}"
    )

    if val_metrics["f1_macro"] > best_f1:
        best_f1 = val_metrics["f1_macro"]
        torch.save(model.state_dict(), f"{SAVE_DIR}/best_model.pt")
        print(f"  -> checkpoint salvo (F1={best_f1:.4f})")

## 5. Resultado

Confirma que o checkpoint foi salvo e exibe as métricas finais. O arquivo `best_model.pt` contém apenas os pesos (não a arquitetura) — para carregar depois:

```python
model = HotelReviewClassifier()
model.load_state_dict(torch.load("models/saved/best_model.pt"))
```

## Avaliação detalhada do checkpoint

Carrega o melhor modelo salvo e exibe F1 por classe — sem retreinar.

In [3]:
from src.models.classifier import HotelReviewClassifier
from src.models.trainer import eval_epoch

eval_model = HotelReviewClassifier().to(DEVICE)
eval_model.load_state_dict(torch.load(f"{SAVE_DIR}/best_model.pt", map_location=DEVICE))

metrics = eval_epoch(eval_model, val_loader, DEVICE)
print(f"F1 macro: {metrics['f1_macro']:.4f}  |  MAE: {metrics['mae']:.4f} (escala 1-5: {metrics['mae']*4:.2f})")
print()
print(metrics["report"])

Loading weights: 100%|██████████| 199/199 [00:00<00:00, 2927.98it/s]
XLMRobertaModel LOAD REPORT from: xlm-roberta-base
Key                       | Status     |  | 
--------------------------+------------+--+-
lm_head.dense.bias        | UNEXPECTED |  | 
lm_head.layer_norm.bias   | UNEXPECTED |  | 
lm_head.dense.weight      | UNEXPECTED |  | 
lm_head.bias              | UNEXPECTED |  | 
lm_head.layer_norm.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


F1 macro: 0.7434  |  MAE: 0.1098 (escala 1-5: 0.44)

              precision    recall  f1-score   support

    negativo       0.68      0.69      0.69      1388
      neutro       0.58      0.68      0.62      2933
    positivo       0.94      0.90      0.92     10943

    accuracy                           0.84     15264
   macro avg       0.73      0.76      0.74     15264
weighted avg       0.85      0.84      0.84     15264



In [ ]:
import os

print(f"Treinamento concluido.")
print(f"Melhor F1 macro: {best_f1:.4f}  (meta: >= 0.80)")

checkpoint_path = f"{SAVE_DIR}/best_model.pt"
if os.path.exists(checkpoint_path):
    size_mb = os.path.getsize(checkpoint_path) / 1024 / 1024
    print(f"Checkpoint salvo em: {checkpoint_path}  ({size_mb:.1f} MB)")
else:
    print("AVISO: checkpoint nao encontrado.")

---

## Benchmark (Opcional)

Roda um treino rápido com 500 reviews para estimar o tempo de um epoch completo. Execute apenas para medir performance — não faz parte do treino real.

> Resultado obtido na RTX 3050: **~54 min/epoch** · Na CPU: ~194 min/epoch

In [ ]:
import time
import torch
from torch.utils.data import DataLoader, Subset
from torch.optim import AdamW
from torch.optim.lr_scheduler import OneCycleLR
from src.data.dataset import ReviewDataset
from src.models.classifier import HotelReviewClassifier
from src.models.trainer import MultiTaskLoss, train_epoch

CSV_PATH = "../data/processed/reviews_labeled.csv"
DEVICE   = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Device: {DEVICE}")

ds     = ReviewDataset(CSV_PATH)
subset = Subset(ds, indices=range(500))
loader = DataLoader(subset, batch_size=32, shuffle=True)

model     = HotelReviewClassifier().to(DEVICE)
criterion = MultiTaskLoss().to(DEVICE)
optimizer = AdamW(model.parameters(), lr=2e-5, weight_decay=0.01)
scheduler = OneCycleLR(optimizer, max_lr=2e-5, steps_per_epoch=len(loader), epochs=1)

print("Rodando benchmark com 500 reviews...")
t0 = time.time()
train_epoch(model, loader, optimizer, criterion, DEVICE, scheduler)
elapsed = time.time() - t0

total_batches_full = 1909
batches_benchmark  = len(loader)
time_per_batch     = elapsed / batches_benchmark
estimated_epoch    = time_per_batch * total_batches_full / 60

print(f"\nBenchmark: {batches_benchmark} batches em {elapsed:.1f}s")
print(f"Tempo por batch: {time_per_batch:.2f}s")
print(f"Estimativa 1 epoch completo: {estimated_epoch:.1f} minutos")